### ISSUE TITLE
[Run Functional Tests] Deployment Items specified in .TestSettings files are not deployed to temp execution locations

### ISSUE THREAD

**Issue by Ryanman** [Microsoft/vsts-tasks#1638]
Team,

I'm still a bit confused as to why mstest has to copy my deployed test files to yet another directory when executing in general, but I've noticed that the execution of this task does not perform the same as it does locally.

The file I specify in the build configuration screen is being correctly pushed to the member of the Machine Group that's intended to run the test:
![image](https://cloud.githubusercontent.com/assets/1376624/15026290/b6e40820-120b-11e6-8d61-b2d215778fbd.png)

The directory containing these files on the member of the machine group is identical to how my solution looks on my dev box:

![image](https://cloud.githubusercontent.com/assets/1376624/15026356/12d6d202-120c-11e6-9115-87e951965be2.png)

And yet, you can see that no files are deployed, using my kung-fu-super-fast-copy-paste skills from the byzantine App Data folder:

![image](https://cloud.githubusercontent.com/assets/1376624/15026561/f43a6eca-120c-11e6-8ad2-0a553930fa93.png)

What gives here? One of the key components of the .testsettings file is the deployment functionality. I can't tell if the testsettings file is being ignored completely or not. If you need more information regarding this ticket please don't hesitate to let me know.

**Comment by bryanmacfarlane** [comment]
Is the context using one of the tasks in this repo?

**Comment by Ryanman** [comment]
Yes - this is using the "Run Functional Tests" task as part of a new build definition.

![image](https://cloud.githubusercontent.com/assets/1376624/15438407/0348316a-1e9a-11e6-9926-2bc1ea08e485.png)

I'm wondering if this is a difference in how the test agent uses .testsettings files to execute, or if I've somehow done something wrong when specifying the location of deployment items. I'm tempted to think it's not the latter because of the what I mentioned: That the directories and artifacts deployed to the agent are identical in structure to how they appear locally/in source control and the deployment items are moved correctly when executing locally.

Thanks for getting back to me.

**Comment by Ryanman** [comment]
@atinb I'm currently working with an on-prem instance of TFS. I'm more than willing to retest once the fix is merged into a release there - I can have DevOPS push the latest and greatest any time. Otherwise I'm not sure if I'll be able to replicate it completely on VSTS since the agents I have at my disposal are going to in our intranet etc.

Thanks for following up. If I can provide any clarification with our existing 2015 R2.1 instance don't hesitate to let me know.

**Comment by atinb** [comment]
Issue 1618 repros when the "Run in Parallel" check box is checked along with a testsettings file. Is this true for you? If so, can you uncheck the box and try? Thanks!

**Comment by Ryanman** [comment]
One important distinction and difference between Issue #1618 and this issue is that I'm executing CodedUI tests using the Run Functional Tests task rather than executing unit tests with the VS Test task. For that reason I obviously don't have the "Run in Parallel" checkbox.

I did have one specified parameter, though - a filter on the Test Names to execute in the assembly. I tried removing that and queuing a new build. The deployment items I specified in .testsettings were not deployed. Here's a screenshot of the relevant portion of the build definition:
![image](https://cloud.githubusercontent.com/assets/1376624/16276910/7fd12158-387e-11e6-929d-3639ea264b2b.png)

**Comment by prawalagarwal** [comment]
Hi @Ryanman ,
I am looking into this issue now. Can you give me some more details? Are you seeing this continuously? Are your tests failing because of these missing files? I will try and reproduce the issue locally.

**Comment by Ryanman** [comment]
Hi - I am seeing this issue continuously. Whenever I queue a build definition for this solution and then Run Functional Tests, nothing is deployed to the strange, random test run folder in App Data beyond:

- The two necessary .dlls
- The config file (named the same as the dll with the Test Methods)
- AgentRestart.dat

My tests don't currently depend on items specified in .testsettings but they may in the future.

**Comment by prawalagarwal** [comment]
Hi Ryan,

The issue is that when tests are remotely, the working directory is by default %temp% and it searches for the files there. For example try manually copying your "SolutionItems" folder to %temp% and run the tests. You will see the files inside getting deployed.

One workaround is that you specify complete path to where you are copying these artifacts in test machines in your testsettings file. Other is to specify the working directory in your testsettings file (check [here](http://stackoverflow.com/questions/2908235/how-to-specify-the-location-for-the-unit-test-results-in-vs-2010)). I hope this helps.

**Comment by prawalagarwal** [comment]
Hi Ryan,
Were you able to resolve the issue?

**Comment by Ryanman** [comment]
@prawalagarwal Adding the necessary artifacts to %temp% did allow them to be deployed along with the test.

Also, thank you for the tip regarding how to specify the working directory in the testsettings file.

This will work in the strictest sense - I'll deploy my artifacts to the test agent, and specify the same spot in TestSettings. But this means that I can't have different folder structures between my dev box and when I'm executing tests on an agent. 

However, I still believe this is an issue that should be rectified. If you have any reason as to why the Automated Testing task (and by extension mstest.exe?) 

- copies and pastes the .dlls and test settings file from deployed directory to a working directory
- THEN attempts to deploy necessary testing artifacts from the working directory... to the working directory 

I'd love to hear it. I may be missing something.

**Comment by atinb** [comment]
We are considering our default working directory to be %temp% instead of the folder specified in the task. Opened a bug to track on our side.

**Comment by prawalagarwal** [comment]
This is issue is now fixed. Removing Dev15 tag. Will close once verified.

**Comment by prawalagarwal** [comment]
Atin will document the usage.
